File: EncryptionService.ts
import crypto from "crypto";

/**
 * Interface defining the contract for encryption and decryption operations.
 * Provides both synchronous and asynchronous methods with optional timeout and cancellation support.
 */
interface IEncryptionService {
  encrypt(plainText: string): string;
  encryptAsync(plainText: string): Promise<string>;
  encryptWithSignal(
    plainText: string,
    signal?: AbortSignal | null
  ): Promise<string>;
  encryptWithTimeout(plainText: string, timeoutMs: number): Promise<string>;
  decrypt(base64CipherText: string): string;
  decryptAsync(base64CipherText: string): Promise<string>;
  decryptWithSignal(
    base64CipherText: string,
    signal?: AbortSignal | null
  ): Promise<string>;
  decryptWithTimeout(
    base64CipherText: string,
    timeoutMs: number
  ): Promise<string>;
  dispose(): void;
}

/**
 * Implements secure encryption and decryption services using AES-256 in CBC mode with HMAC-SHA256.
 * Features include:
 * - AES-256 encryption in CBC mode
 * - Cryptographically secure random IV generation
 * - HMAC-SHA256 for integrity verification
 * - Secure key management with proper disposal
 * - Support for both sync and async operations
 * - Timeout and cancellation support
 */
class EncryptionService implements IEncryptionService {
  // Constants
  private static readonly IV_SIZE: number = 16;
  private static readonly HMAC_SIZE: number = 32;
  private static readonly KEY_SIZE: number = 32; // 256-bit
  private static readonly DEFAULT_TIMEOUT: number = 30000; // 30 seconds

  private readonly _encryptionKey: Buffer;
  private _isDisposed: boolean;

  constructor(base64Key: string) {
    this._validateEncryptionKey(base64Key);
    this._encryptionKey = Buffer.from(base64Key, "base64");
    this._isDisposed = false;
  }

  /**
   * Encrypts the given plain text synchronously using the default timeout.
   * @param plainText - The text to encrypt
   * @returns Base64 encoded encrypted string
   * @throws Error if the service is disposed or input is invalid
   */
  public encrypt(plainText: string): string {
    this._validateInputData(plainText);
    return this._performEncryption(plainText);
  }

  /**
   * Decrypts the given Base64 encoded cipher text synchronously.
   * @param base64CipherText - The Base64 encoded cipher text to decrypt
   * @returns Decrypted plain text
   * @throws Error if the service is disposed, input is invalid, or integrity check fails
   */
  public decrypt(base64CipherText: string): string {
    this._validateInputData(base64CipherText);
    return this._performDecryption(base64CipherText);
  }

  /**
   * Encrypts the given plain text asynchronously using the default timeout.
   * @param plainText - The text to encrypt
   * @returns Promise resolving to Base64 encoded encrypted string
   * @throws Error if the service is disposed or input is invalid
   */
  public async encryptAsync(plainText: string): Promise<string> {
    return await this.encryptWithTimeout(
      plainText,
      EncryptionService.DEFAULT_TIMEOUT
    );
  }

  /**
   * Encrypts the given plain text with cancellation support.
   * @param plainText - The text to encrypt
   * @param signal - Optional AbortSignal for cancellation
   * @returns Promise resolving to Base64 encoded encrypted string
   * @throws Error if the operation is cancelled, service is disposed, or input is invalid
   */
  public encryptWithSignal(
    plainText: string,
    signal: AbortSignal | null = null
  ): Promise<string> {
    this._validateInputData(plainText);
    return this._performEncryptionAsync(plainText, signal);
  }

  /**
   * Encrypts the given plain text with a specified timeout.
   * @param plainText - The text to encrypt
   * @param timeoutMs - Maximum time in milliseconds allowed for the operation
   * @returns Promise resolving to Base64 encoded encrypted string
   * @throws Error if the operation times out, service is disposed, or input is invalid
   */
  public async encryptWithTimeout(
    plainText: string,
    timeoutMs: number
  ): Promise<string> {
    this._validateInputData(plainText);
    const controller = new AbortController();
    const timeout = setTimeout(() => controller.abort(), timeoutMs);

    try {
      return await this._performEncryptionAsync(plainText, controller.signal);
    } finally {
      clearTimeout(timeout);
    }
  }

  /**
   * Decrypts the given Base64 encoded cipher text asynchronously.
   * @param base64CipherText - The Base64 encoded cipher text to decrypt
   * @returns Promise resolving to decrypted plain text
   * @throws Error if the service is disposed, input is invalid, or integrity check fails
   */
  public async decryptAsync(base64CipherText: string): Promise<string> {
    return await this.decryptWithTimeout(
      base64CipherText,
      EncryptionService.DEFAULT_TIMEOUT
    );
  }

  /**
   * Decrypts the given Base64 encoded cipher text with cancellation support.
   * @param base64CipherText - The Base64 encoded cipher text to decrypt
   * @param signal - Optional AbortSignal for cancellation
   * @returns Promise resolving to decrypted plain text
   * @throws Error if cancelled, service is disposed, input is invalid, or integrity check fails
   */
  public decryptWithSignal(
    base64CipherText: string,
    signal: AbortSignal | null = null
  ): Promise<string> {
    this._validateInputData(base64CipherText);
    return this._performDecryptionAsync(base64CipherText, signal);
  }

  /**
   * Decrypts the given Base64 encoded cipher text with a specified timeout.
   * @param base64CipherText - The Base64 encoded cipher text to decrypt
   * @param timeoutMs - Maximum time in milliseconds allowed for the operation
   * @returns Promise resolving to decrypted plain text
   * @throws Error if times out, service is disposed, input is invalid, or integrity check fails
   */
  public async decryptWithTimeout(
    base64CipherText: string,
    timeoutMs: number
  ): Promise<string> {
    this._validateInputData(base64CipherText);
    const controller = new AbortController();
    const timeout = setTimeout(() => controller.abort(), timeoutMs);

    try {
      return await this._performDecryptionAsync(
        base64CipherText,
        controller.signal
      );
    } finally {
      clearTimeout(timeout);
    }
  }

  /**
   * Disposes the encryption service by securely clearing the encryption key from memory.
   */
  public dispose(): void {
    if (!this._isDisposed) {
      this._isDisposed = true;
      if (this._encryptionKey) {
        this._encryptionKey.fill(0); // Clear the key from memory
      }
    }
  }

  // Private Methods
  private _performEncryption(plainText: string): string {
    // Generate IV
    const iv = crypto.randomBytes(EncryptionService.IV_SIZE);

    // Convert text to bytes with UTF-8 encoding
    const plainBytes = Buffer.from(plainText, "utf8");

    // Encrypt the data
    const encryptedData = this._encryptDataWithIV(plainBytes, iv, null);

    // Combine IV and encrypted data
    const combined = Buffer.concat([iv, encryptedData]);

    // Calculate HMAC
    const hmac = this._calculateHMAC(combined);

    // Combine HMAC + IV + encrypted data
    const result = Buffer.concat([hmac, combined]);

    // Return as base64
    return result.toString("base64");
  }

  private async _performEncryptionAsync(
    plainText: string,
    signal: AbortSignal | null
  ): Promise<string> {
    // Generate IV
    const iv = crypto.randomBytes(EncryptionService.IV_SIZE);

    // Convert text to bytes with UTF-8 encoding
    const plainBytes = Buffer.from(plainText, "utf8");

    // Encrypt the data
    const encryptedData = this._encryptDataWithIV(plainBytes, iv, signal);

    // Combine IV and encrypted data
    const combined = Buffer.concat([iv, encryptedData]);

    // Calculate HMAC
    const hmac = this._calculateHMAC(combined);

    // Combine HMAC + IV + encrypted data
    const result = Buffer.concat([hmac, combined]);

    // Return as base64
    return result.toString("base64");
  }

  private _performDecryption(base64CipherText: string): string {
    const cipherData = this._parseAndValidateBase64Data(base64CipherText);

    const hmac = cipherData.slice(0, EncryptionService.HMAC_SIZE);
    const encryptedData = cipherData.slice(EncryptionService.HMAC_SIZE);

    this._verifyHMAC(encryptedData, hmac);

    const iv = encryptedData.slice(0, EncryptionService.IV_SIZE);
    const cipherText = encryptedData.slice(EncryptionService.IV_SIZE);

    const result = this._decryptDataWithIV(cipherText, iv);
    cipherData.fill(0); // Clear sensitive data
    return result;
  }

  private async _performDecryptionAsync(
    base64CipherText: string,
    signal: AbortSignal | null
  ): Promise<string> {
    const cipherData = this._parseAndValidateBase64Data(base64CipherText);

    const hmac = cipherData.slice(0, EncryptionService.HMAC_SIZE);
    const encryptedData = cipherData.slice(EncryptionService.HMAC_SIZE);

    this._verifyHMAC(encryptedData, hmac);

    const iv = encryptedData.slice(0, EncryptionService.IV_SIZE);
    const cipherText = encryptedData.slice(EncryptionService.IV_SIZE);

    const result = await this._decryptDataWithIVAsync(cipherText, iv, signal);
    cipherData.fill(0); // Clear sensitive data
    return result;
  }

  private _encryptDataWithIV(
    plainBytes: Buffer,
    iv: Buffer,
    signal: AbortSignal | null
  ): Buffer {
    if (signal?.aborted) {
      throw new Error("Operation cancelled");
    }

    const cipher = crypto.createCipheriv(
      "aes-256-cbc",
      this._encryptionKey,
      iv
    );
    return Buffer.concat([cipher.update(plainBytes), cipher.final()]);
  }

  private async _decryptDataWithIVAsync(
    cipherText: Buffer,
    iv: Buffer,
    signal: AbortSignal | null
  ): Promise<string> {
    if (signal?.aborted) {
      throw new Error("Operation cancelled");
    }

    const decipher = crypto.createDecipheriv(
      "aes-256-cbc",
      this._encryptionKey,
      iv
    );
    const decrypted = Buffer.concat([
      decipher.update(cipherText),
      decipher.final(),
    ]);

    return decrypted.toString("utf8");
  }

  private _decryptDataWithIV(cipherText: Buffer, iv: Buffer): string {
    const decipher = crypto.createDecipheriv(
      "aes-256-cbc",
      this._encryptionKey,
      iv
    );
    const decrypted = Buffer.concat([
      decipher.update(cipherText),
      decipher.final(),
    ]);

    return decrypted.toString("utf8");
  }

  private _calculateHMAC(data: Buffer): Buffer {
    const hmac = crypto.createHmac("sha256", this._encryptionKey);
    hmac.update(data);
    return hmac.digest();
  }

  private _verifyHMAC(data: Buffer, expectedHmac: Buffer): void {
    const actualHmac = this._calculateHMAC(data);
    if (!crypto.timingSafeEqual(actualHmac, expectedHmac)) {
      throw new Error(
        "Data verification failed! Data integrity might be compromised."
      );
    }
  }

  private _parseAndValidateBase64Data(base64Text: string): Buffer {
    let data: Buffer;
    try {
      data = Buffer.from(base64Text, "base64");
    } catch {
      throw new Error("Invalid Base64 format for encrypted text!");
    }

    if (data.length < EncryptionService.HMAC_SIZE + EncryptionService.IV_SIZE) {
      data.fill(0);
      throw new Error("Invalid encrypted text length!");
    }

    return data;
  }

  private _validateEncryptionKey(base64Key: string): void {
    if (!base64Key) {
      throw new Error("Encryption key cannot be empty!");
    }

    let key: Buffer;
    try {
      key = Buffer.from(base64Key, "base64");
    } catch {
      throw new Error("Invalid Base64 format for encryption key!");
    }

    const isValidLength = key.length === EncryptionService.KEY_SIZE;
    key.fill(0);

    if (!isValidLength) {
      throw new Error(
        `Key length must be ${EncryptionService.KEY_SIZE} bytes!`
      );
    }
  }

  private _validateInputData(input: string): void {
    if (this._isDisposed) {
      throw new Error("Cannot access a disposed EncryptionService.");
    }
    if (!input) {
      throw new Error("Input data cannot be empty!");
    }
  }
}

// Helper functions
/**
 * Helper function to encrypt text synchronously with automatic resource cleanup.
 * @param plainText - The text to encrypt
 * @param base64Key - The Base64 encoded encryption key
 * @returns Base64 encoded encrypted string
 */
function encrypt(plainText: string, base64Key: string): string {
  const encryptor = new EncryptionService(base64Key);
  try {
    return encryptor.encrypt(plainText);
  } finally {
    encryptor.dispose();
  }
}

/**
 * Helper function to decrypt text synchronously with automatic resource cleanup.
 * @param encryptedValue - The Base64 encoded encrypted text
 * @param base64Key - The Base64 encoded encryption key
 * @returns Decrypted plain text
 */
function decrypt(encryptedValue: string, base64Key: string): string {
  const decryptor = new EncryptionService(base64Key);
  try {
    return decryptor.decrypt(encryptedValue);
  } finally {
    decryptor.dispose();
  }
}

/**
 * Helper function to encrypt text asynchronously with automatic resource cleanup.
 * @param plainText - The text to encrypt
 * @param base64Key - The Base64 encoded encryption key
 * @returns Promise resolving to Base64 encoded encrypted string
 */
async function encryptAsync(
  plainText: string,
  base64Key: string
): Promise<string> {
  const encryptor = new EncryptionService(base64Key);
  try {
    return await encryptor.encryptAsync(plainText);
  } finally {
    encryptor.dispose();
  }
}

/**
 * Helper function to decrypt text asynchronously with automatic resource cleanup.
 * @param encryptedValue - The Base64 encoded encrypted text
 * @param base64Key - The Base64 encoded encryption key
 * @returns Promise resolving to decrypted plain text
 */
async function decryptAsync(
  encryptedValue: string,
  base64Key: string
): Promise<string> {
  const decryptor = new EncryptionService(base64Key);
  try {
    return await decryptor.decryptAsync(encryptedValue);
  } finally {
    decryptor.dispose();
  }
}

/**
 * Helper function to encrypt text with cancellation support and automatic resource cleanup.
 * @param plainText - The text to encrypt
 * @param base64Key - The Base64 encoded encryption key
 * @param signal - Optional AbortSignal for cancellation
 * @returns Promise resolving to Base64 encoded encrypted string
 */
async function encryptWithSignal(
  plainText: string,
  base64Key: string,
  signal: AbortSignal | null = null
): Promise<string> {
  const encryptor = new EncryptionService(base64Key);
  try {
    return await encryptor.encryptWithSignal(plainText, signal);
  } finally {
    encryptor.dispose();
  }
}

/**
 * Helper function to decrypt text with cancellation support and automatic resource cleanup.
 * @param encryptedValue - The Base64 encoded encrypted text
 * @param base64Key - The Base64 encoded encryption key
 * @param signal - Optional AbortSignal for cancellation
 * @returns Promise resolving to decrypted plain text
 */
async function decryptWithSignal(
  encryptedValue: string,
  base64Key: string,
  signal: AbortSignal | null = null
): Promise<string> {
  const decryptor = new EncryptionService(base64Key);
  try {
    return await decryptor.decryptWithSignal(encryptedValue, signal);
  } finally {
    decryptor.dispose();
  }
}

export {
  IEncryptionService,
  EncryptionService,
  encrypt,
  decrypt,
  encryptAsync,
  decryptAsync,
  encryptWithSignal,
  decryptWithSignal,
};

File: EncryptionService.cs
using System.Security.Cryptography;
using System.Text;
using System.Buffers;
using System.Runtime.CompilerServices;
using Microsoft.Extensions.ObjectPool;

public interface IEncryptionService : IDisposable
{
    /// <summary>
    /// Encrypts the plain text using default timeout.
    /// </summary>
    /// <param name="plainText">The text to encrypt</param>
    /// <returns>Base64 encoded encrypted string</returns>
    string Encrypt(string plainText);

    /// <summary>
    /// Encrypts the plain text asynchronously.
    /// </summary>
    /// <param name="plainText">The text to encrypt</param>
    /// <param name="cancellationToken">Cancellation token to cancel the operation</param>
    /// <returns>Base64 encoded encrypted string</returns>
    Task<string> EncryptAsync(string plainText, CancellationToken cancellationToken = default);

    /// <summary>
    /// Encrypts the plain text with a specified timeout.
    /// </summary>
    /// <param name="plainText">The text to encrypt</param>
    /// <param name="timeout">Maximum time allowed for the operation</param>
    /// <returns>Base64 encoded encrypted string</returns>
    string Encrypt(string plainText, TimeSpan timeout);

    /// <summary>
    /// Decrypts the cipher text using default timeout.
    /// </summary>
    /// <param name="base64CipherText">The Base64 encoded cipher text to decrypt</param>
    /// <returns>Decrypted plain text</returns>
    string Decrypt(string base64CipherText);

    /// <summary>
    /// Decrypts the cipher text asynchronously.
    /// </summary>
    /// <param name="base64CipherText">The Base64 encoded cipher text to decrypt</param>
    /// <param name="cancellationToken">Cancellation token to cancel the operation</param>
    /// <returns>Decrypted plain text</returns>
    Task<string> DecryptAsync(string base64CipherText, CancellationToken cancellationToken = default);

    /// <summary>
    /// Decrypts the cipher text with a specified timeout.
    /// </summary>
    /// <param name="base64CipherText">The Base64 encoded cipher text to decrypt</param>
    /// <param name="timeout">Maximum time allowed for the operation</param>
    /// <returns>Decrypted plain text</returns>
    string Decrypt(string base64CipherText, TimeSpan timeout);
}

/// <summary>
/// Provides secure encryption and decryption services using AES-256 in CBC mode with HMAC-SHA256 for integrity verification.
/// This service implements object pooling for better performance and proper resource management.
/// </summary>
/// <remarks>
/// The service uses the following security features:
/// - AES-256 encryption in CBC mode with PKCS7 padding
/// - Cryptographically secure random IV generation
/// - HMAC-SHA256 for integrity verification
/// - Secure key management with proper disposal
/// - Object pooling for better performance
/// </remarks>
public sealed class EncryptionService : IEncryptionService
{
    #region Constants
    private const int IvSize = 16;
    private const int HmacSize = 32;
    private const int KeySize = 32; // 256-bit
    private const int MinBufferSize = 1024; // 1KB
    private const int MaxBufferSize = 1024 * 1024; // 1MB
    private const int DefaultBufferSize = 4096; // 4KB
    private const int PoolSize = 32; // Maximum number of pooled objects
    private static readonly TimeSpan DefaultTimeout = TimeSpan.FromSeconds(30);
    #endregion

    #region Pool Policies
    private sealed class AesPoolPolicy : PooledObjectPolicy<Aes>
    {
        public override Aes Create()
        {
            var aes = Aes.Create();
            aes.Mode = CipherMode.CBC;
            aes.Padding = PaddingMode.PKCS7;
            aes.KeySize = 256;
            return aes;
        }

        public override bool Return(Aes obj)
        {
            if (obj?.Key == null)
                return false;

            Array.Clear(obj.Key, 0, obj.Key.Length);
            return true;
        }
    }

    private sealed class HmacPoolPolicy : PooledObjectPolicy<HMACSHA256>
    {
        private readonly byte[] _key;

        public HmacPoolPolicy(byte[] key) => _key = key;

        public override HMACSHA256 Create() => new HMACSHA256(_key);

        public override bool Return(HMACSHA256 obj)
        {
            if (obj?.Key == null)
                return false;

            Array.Clear(obj.Key, 0, obj.Key.Length);
            return true;
        }
    }
    #endregion

    #region Fields
    private readonly byte[] _key;
    private readonly DefaultObjectPool<Aes> _aesPool;
    private readonly DefaultObjectPool<HMACSHA256> _hmacPool;
    private bool _disposed;
    #endregion

    #region Constructor & Dispose
    /// <summary>
    /// Initializes a new instance of the EncryptionService with the specified encryption key.
    /// </summary>
    /// <param name="base64Key">The Base64 encoded 256-bit encryption key</param>
    /// <exception cref="ArgumentException">Thrown when the key is invalid or has incorrect length</exception>
    public EncryptionService(string base64Key)
    {
        ValidateKey(base64Key);
        _key = Convert.FromBase64String(base64Key);
        _aesPool = new DefaultObjectPool<Aes>(new AesPoolPolicy(), PoolSize);
        _hmacPool = new DefaultObjectPool<HMACSHA256>(new HmacPoolPolicy(_key), PoolSize);
    }

    /// <summary>
    /// Disposes the encryption service and securely clears the encryption key from memory.
    /// </summary>
    public void Dispose()
    {
        if (!_disposed)
        {
            _disposed = true;
            Array.Clear(_key, 0, _key.Length);
            GC.SuppressFinalize(this);
        }
    }
    #endregion

    #region Public Methods
    [MethodImpl(MethodImplOptions.AggressiveInlining)]
    public string Encrypt(string plainText) => Encrypt(plainText, DefaultTimeout);

    public Task<string> EncryptAsync(string plainText, CancellationToken cancellationToken = default)
    {
        ValidateInput(plainText);
        return Task.FromResult(EncryptInternal(plainText.AsSpan(), cancellationToken));
    }

    public string Encrypt(string plainText, TimeSpan timeout)
    {
        ValidateInput(plainText);
        using var cts = new CancellationTokenSource(timeout);
        return EncryptInternal(plainText.AsSpan(), cts.Token);
    }

    [MethodImpl(MethodImplOptions.AggressiveInlining)]
    public string Decrypt(string base64CipherText) => Decrypt(base64CipherText, DefaultTimeout);

    public Task<string> DecryptAsync(string base64CipherText, CancellationToken cancellationToken = default)
    {
        ValidateInput(base64CipherText);
        return Task.FromResult(DecryptInternal(base64CipherText, cancellationToken));
    }

    public string Decrypt(string base64CipherText, TimeSpan timeout)
    {
        ValidateInput(base64CipherText);
        using var cts = new CancellationTokenSource(timeout);
        return DecryptInternal(base64CipherText, cts.Token);
    }
    #endregion

    #region Private Methods
    private string EncryptInternal(ReadOnlySpan<char> plainText, CancellationToken cancellationToken)
    {
        // Get IV
        Span<byte> iv = stackalloc byte[IvSize];
        RandomNumberGenerator.Fill(iv);

        // Get plain bytes with minimal allocations
        int maxPlainTextBytes = Encoding.UTF8.GetMaxByteCount(plainText.Length);
        byte[] rentedBuffer = ArrayPool<byte>.Shared.Rent(maxPlainTextBytes);
        int actualPlainTextBytes = 0;

        actualPlainTextBytes = Encoding.UTF8.GetBytes(plainText, rentedBuffer);
        byte[] encryptedData = EncryptData(rentedBuffer.AsSpan(0, actualPlainTextBytes), iv, cancellationToken);
        Array.Clear(rentedBuffer, 0, actualPlainTextBytes);
        ArrayPool<byte>.Shared.Return(rentedBuffer);

        // Combine IV and encrypted data on stack
        Span<byte> combined = stackalloc byte[iv.Length + encryptedData.Length];
        iv.CopyTo(combined);
        encryptedData.CopyTo(combined[iv.Length..]);

        // Calculate HMAC
        byte[] hmac = ComputeHmac(combined);

        // Prepare final result with minimal allocations
        byte[] result = new byte[HmacSize + combined.Length];
        hmac.CopyTo(result.AsSpan(0, HmacSize));
        combined.CopyTo(result.AsSpan(HmacSize));

        return Convert.ToBase64String(result);
    }

    private string DecryptInternal(string base64CipherText, CancellationToken cancellationToken)
    {
        byte[] cipherData = ParseAndValidateBase64(base64CipherText);
        ReadOnlySpan<byte> hmac = cipherData.AsSpan(0, HmacSize);
        ReadOnlySpan<byte> encryptedData = cipherData.AsSpan(HmacSize);

        ValidateHmac(encryptedData, hmac);

        ReadOnlySpan<byte> iv = encryptedData[..IvSize];
        ReadOnlySpan<byte> cipherText = encryptedData[IvSize..];

        string result = DecryptData(cipherText, iv, cancellationToken);
        Array.Clear(cipherData, 0, cipherData.Length);
        return result;
    }

    /// <summary>
    /// Encrypts the given plain text data with the specified initialization vector.
    /// </summary>
    /// <param name="plainBytes">The plain text data to encrypt</param>
    /// <param name="iv">The initialization vector for encryption</param>
    /// <param name="cancellationToken">Token to monitor for cancellation requests</param>
    /// <returns>The encrypted data</returns>
    [MethodImpl(MethodImplOptions.AggressiveInlining)]
    private byte[] EncryptData(ReadOnlySpan<byte> plainBytes, ReadOnlySpan<byte> iv, CancellationToken cancellationToken)
    {
        var aes = _aesPool.Get();
        if (aes is null) return Array.Empty<byte>();

        aes.Key = _key;
        aes.IV = iv.ToArray();

        using var encryptor = aes.CreateEncryptor();
        int maxEncryptedSize = plainBytes.Length + (aes.BlockSize / 8);
        int bufferSize = Math.Min(Math.Max(maxEncryptedSize, MinBufferSize), MaxBufferSize);

        using var ms = new MemoryStream(bufferSize);
        using (var cs = new CryptoStream(ms, encryptor, CryptoStreamMode.Write, leaveOpen: true))
        {
            cs.Write(plainBytes);
            cs.FlushFinalBlock();
        }

        cancellationToken.ThrowIfCancellationRequested();
        _aesPool.Return(aes);
        return ms.ToArray();
    }

    /// <summary>
    /// Decrypts the given cipher text using the specified initialization vector.
    /// </summary>
    /// <param name="cipherText">The encrypted data to decrypt</param>
    /// <param name="iv">The initialization vector used for encryption</param>
    /// <param name="cancellationToken">Token to monitor for cancellation requests</param>
    /// <returns>The decrypted data as a string</returns>
    private string DecryptData(ReadOnlySpan<byte> cipherText, ReadOnlySpan<byte> iv, CancellationToken cancellationToken)
    {
        var aes = _aesPool.Get();
        if (aes is null) return string.Empty;

        aes.Key = _key;
        aes.IV = iv.ToArray();

        using var decryptor = aes.CreateDecryptor();
        int bufferSize = Math.Min(Math.Max(cipherText.Length, MinBufferSize), MaxBufferSize);
        using var outputBuffer = new MemoryStream(bufferSize);

        var rentedBuffer = ArrayPool<byte>.Shared.Rent(bufferSize);
        using (var ms = new MemoryStream(cipherText.ToArray(), writable: false))
        using (var cs = new CryptoStream(ms, decryptor, CryptoStreamMode.Read))
        {
            int bytesRead;
            while ((bytesRead = cs.Read(rentedBuffer, 0, bufferSize)) > 0)
            {
                cancellationToken.ThrowIfCancellationRequested();
                outputBuffer.Write(rentedBuffer, 0, bytesRead);
            }
        }

        Array.Clear(rentedBuffer, 0, rentedBuffer.Length);
        ArrayPool<byte>.Shared.Return(rentedBuffer);
        _aesPool.Return(aes);

        outputBuffer.Position = 0;
        using var reader = new StreamReader(outputBuffer, Encoding.UTF8, detectEncodingFromByteOrderMarks: false);
        return reader.ReadToEnd();
    }

    /// <summary>
    /// Computes the HMAC-SHA256 for the given data using the encryption key.
    /// </summary>
    /// <param name="data">The data to compute HMAC for</param>
    /// <returns>The computed HMAC</returns>
    private byte[] ComputeHmac(ReadOnlySpan<byte> data)
    {
        var hmac = _hmacPool.Get();
        if (hmac is null) return Array.Empty<byte>();

        var result = hmac.ComputeHash(data.ToArray());
        _hmacPool.Return(hmac);
        return result;
    }

    /// <summary>
    /// Validates the HMAC of the encrypted data to ensure data integrity.
    /// </summary>
    /// <param name="data">The encrypted data to validate</param>
    /// <param name="expectedHmac">The expected HMAC value</param>
    /// <exception cref="CryptographicException">Thrown when HMAC validation fails</exception>
    private void ValidateHmac(ReadOnlySpan<byte> data, ReadOnlySpan<byte> expectedHmac)
    {
        var hmac = _hmacPool.Get();
        if (hmac is null) return;

        var actualHmac = hmac.ComputeHash(data.ToArray());
        var isValid = CryptographicOperations.FixedTimeEquals(actualHmac, expectedHmac.ToArray());

        Array.Clear(actualHmac, 0, actualHmac.Length);
        _hmacPool.Return(hmac);

        if (!isValid)
        {
            throw new CryptographicException("Data verification failed! Data integrity might be compromised.");
        }
    }

    /// <summary>
    /// Parses and validates the Base64 encoded cipher text.
    /// </summary>
    /// <param name="base64Text">The Base64 encoded cipher text to parse</param>
    /// <returns>The decoded cipher data</returns>
    /// <exception cref="ArgumentException">Thrown when the input is not valid Base64</exception>
    /// <exception cref="CryptographicException">Thrown when the decoded data length is invalid</exception>
    private static byte[] ParseAndValidateBase64(string base64Text)
    {
        if (!Convert.TryFromBase64String(base64Text, new byte[base64Text.Length], out _))
        {
            throw new ArgumentException("Invalid Base64 format for encrypted text!");
        }

        var data = Convert.FromBase64String(base64Text);
        if (data.Length < HmacSize + IvSize)
        {
            Array.Clear(data, 0, data.Length);
            throw new CryptographicException("Invalid encrypted text length!");
        }

        return data;
    }

    /// <summary>
    /// Validates the encryption key format and length.
    /// </summary>
    /// <param name="base64Key">The Base64 encoded key to validate</param>
    /// <exception cref="ArgumentException">Thrown when the key is invalid or has incorrect length</exception>
    private static void ValidateKey(string base64Key)
    {
        ArgumentException.ThrowIfNullOrWhiteSpace(base64Key, nameof(base64Key));

        if (!Convert.TryFromBase64String(base64Key, new byte[base64Key.Length], out _))
        {
            throw new ArgumentException("Invalid Base64 format for encryption key!", nameof(base64Key));
        }

        var key = Convert.FromBase64String(base64Key);
        var isValidLength = key.Length == KeySize;
        Array.Clear(key, 0, key.Length);

        if (!isValidLength)
        {
            throw new ArgumentException($"Key length must be {KeySize} bytes!", nameof(base64Key));
        }
    }

    /// <summary>
    /// Validates the input data and service state.
    /// </summary>
    /// <param name="input">The input data to validate</param>
    /// <exception cref="ObjectDisposedException">Thrown when the service is disposed</exception>
    /// <exception cref="ArgumentException">Thrown when the input is null or empty</exception>
    private void ValidateInput(string input)
    {
        ObjectDisposedException.ThrowIf(_disposed, nameof(EncryptionService));
        ArgumentException.ThrowIfNullOrWhiteSpace(input, nameof(input));
    }
    #endregion
}
